File: referralhero/src/main/java/com/sdk/rh/networking/ReferralNetworkClient.java
package com.sdk.rh.networking;

import android.util.Log;

import androidx.annotation.NonNull;

import com.google.gson.Gson;
import com.google.gson.JsonSyntaxException;
import com.google.gson.reflect.TypeToken;
import com.sdk.rh.PrefHelper;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.Map;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.FormBody;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class ReferralNetworkClient {

    private OkHttpClient client;
    private final Gson gson;

    public ReferralNetworkClient() {
        this.client = new OkHttpClient();
        this.gson = new Gson();
    }

    public void callApiGet(String endpoint, Map<String, String> queryParams, Callback callback) {
        HttpUrl.Builder urlBuilder = HttpUrl.parse(PrefHelper.getAPIBaseUrl() + endpoint).newBuilder();
        if (queryParams != null) {
            for (Map.Entry<String, String> entry : queryParams.entrySet()) {
                urlBuilder.addQueryParameter(entry.getKey(), entry.getValue());
            }
        }
        String url = urlBuilder.build().toString();

        Request request = new Request.Builder()
                .url(url)
                .build();

        Call call = client.newCall(request);
        call.enqueue(callback);
    }

    public <T> void callApiPost(String endpoint, Map<String, String> params, Type responseType, ServerCallback callback) {
        FormBody.Builder formBuilder = new FormBody.Builder();
        if (params != null) {
            for (Map.Entry<String, String> entry : params.entrySet()) {
                formBuilder.add(entry.getKey(), entry.getValue());
            }
        }
        RequestBody requestBody = formBuilder.build();

        HttpUrl.Builder urlBuilder = HttpUrl.parse(PrefHelper.getAPIBaseUrl() + endpoint).newBuilder();
        String url = urlBuilder.build().toString();

        Request request = new Request.Builder()
                .url(url)
                .post(requestBody)
                .build();

        Call call = client.newCall(request);
        call.enqueue(new Callback() {
            @Override
            public void onFailure(@NonNull Call call, @NonNull IOException e) {
                callback.onFailure(call, e);
            }

            @Override
            public void onResponse(@NonNull Call call, @NonNull Response response) {
                try {
                    String responseString = response.body().string();
                    Log.e("responseString",responseString);
                    Type apiResponseType = TypeToken.getParameterized(ApiResponse.class, responseType).getType();
                    ApiResponse parsedResponse = gson.fromJson(responseString, apiResponseType);
                    callback.onResponse(call, parsedResponse);
                } catch (Exception e) {
                    callback.onFailure(call, e);
                }
            }
        });
    }


}

File: referralhero/src/main/java/com/sdk/rh/networking/ReferralNetworkClient.kt
package com.sdk.rh.networking

import android.content.Context
import android.util.Log
import com.google.gson.Gson
import com.sdk.rh.PrefHelper
import com.sdk.rh.RHUtil
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody

class ReferralNetworkClient {
    private val client: OkHttpClient = OkHttpClient()
    private val gson: Gson = Gson()


    /**
    - Created a new function `serverRequestGetAsync` to handle GET requests.
    - Removed the creation of the request body since it's not needed for GET requests.
    - Utilized the `get()` method to specify the request type as GET.
    - Retained the logic for building the URL and adding headers.
    - Used `withContext` and `Dispatchers.IO` for performing the network request asynchronously.
    - Handled the response based on the HTTP status code, returning the parsed response if code is 200, otherwise returning a new `ApiResponse` object with the error details.
     * */
    suspend fun <T> serverRequestGetAsync(
        context: Context, endpoint: String
    ): ApiResponse {

        val urlBuilder = (PrefHelper.aPIBaseUrl + endpoint).toHttpUrlOrNull()?.newBuilder()

        val url = urlBuilder?.build()?.toString()

        Log.e("URL", url.toString())
        val requestBuilder =
            Request.Builder().url(url!!).addHeader("Authorization", RHUtil.readRhKey(context))
                .addHeader("Accept", "application/vnd.referralhero.v1")
                .addHeader("Content-Type", "application/json").get()

        val request = requestBuilder.build()
        return withContext(Dispatchers.IO) {
            val response = client.newCall(request).execute()
            val responseString = response.body?.string()
            val parsedResponse: ApiResponse = gson.fromJson(responseString, ApiResponse::class.java)
            if (response.code == 200) {
                parsedResponse
            } else {
                ApiResponse("error", parsedResponse.message, parsedResponse.code, null, null, 0)
            }
        }
    }

    /**
    - Extracted the creation of the JSON request body into a separate variable for improved readability.
    - Used Kotlin's null-safe operators for building the URL and accessing its string representation.
    - Set the appropriate media type for the request body.
    - Added headers for authorization and accept content type.
    - Used `withContext` and `Dispatchers.IO` for performing the network request asynchronously.
    - Handled the response based on the HTTP status code, returning the parsed response if code is 200, otherwise returning a new `ApiResponse` object with the error details.
     * ***/

    suspend fun <T> serverRequestCallBackAsync(
        context: Context, endpoint: String, referralParams: ReferralParams
    ): ApiResponse {

        val jsonMediaType = "application/json".toMediaTypeOrNull()
        val requestBody: RequestBody = Gson().toJson(referralParams).toRequestBody(jsonMediaType)
        val urlBuilder = (PrefHelper.aPIBaseUrl + endpoint).toHttpUrlOrNull()?.newBuilder()
        val url = urlBuilder?.build()?.toString()

        val requestBuilder =
            Request.Builder().url(url!!).addHeader("Authorization", RHUtil.readRhKey(context))
                .addHeader("Accept", "application/vnd.referralhero.v1").post(requestBody)

        Log.e("URL", url)
        val request = requestBuilder.build()
        return withContext(Dispatchers.IO) {
            val response = client.newCall(request).execute()
            val responseString = response.body?.string()
            val parsedResponse: ApiResponse = gson.fromJson(responseString, ApiResponse::class.java)
            if (response.code == 200) {
                parsedResponse
            } else {
                ApiResponse("error", parsedResponse.message, parsedResponse.code, null, null, 0)
            }

        }
    }

    /**
    - Created a new function `serverRequestDeleteAsync` to handle DELETE requests.
    - Updated the function signature to include the required parameters: `context`, `endpoint`, and optional `queryParams`.
    - Utilized the `delete()` method of the `Request.Builder` class to specify the request type as DELETE.
    - Appended query parameters to the URL if provided using the `addQueryParameter` method.
    - Set the appropriate headers for authorization, accept content type, and content type.
    - Used `withContext` and `Dispatchers.IO` for performing the network request asynchronously.
    - Handled the response based on the HTTP status code, returning the parsed response if code is 200, otherwise returning a new `ApiResponse` object with the error details.
     * ***/
    suspend fun <T> serverRequestDeleteAsync(
        context: Context, endpoint: String
    ): ApiResponse {

        val urlBuilder = (PrefHelper.aPIBaseUrl + endpoint).toHttpUrlOrNull()?.newBuilder()
        val url = urlBuilder?.build()?.toString()

        val requestBuilder =
            Request.Builder().url(url!!).addHeader("Authorization", RHUtil.readRhKey(context))
                .addHeader("Accept", "application/vnd.referralhero.v1")
                .addHeader("Content-Type", "application/json").delete()

        val request = requestBuilder.build()
        return withContext(Dispatchers.IO) {
            val response = client.newCall(request).execute()
            val responseString = response.body?.string()
            val parsedResponse: ApiResponse = gson.fromJson(responseString, ApiResponse::class.java)
            if (response.code == 200) {
                parsedResponse
            } else {
                ApiResponse("error", parsedResponse.message, parsedResponse.code, null, null, 0)
            }
        }
    }

    /**
    - Added a new function 'serverRequestPatchAsync' to handle PATCH requests.
    - Updated the function signature to include the required parameters: `context`, `endpoint`, and `requestBody`.
    - Built the URL using the `aPIBaseUrl` and `endpoint` parameters.
    - Set the appropriate headers for authorization, accept content type, and content type.
    - Utilized the `patch()` method of the `Request.Builder` class to specify the request type as PATCH.
    - Used `withContext` and `Dispatchers.IO` for performing the network request asynchronously.
    - Handled the response based on the HTTP status code, returning the parsed response if the code is 200, otherwise returning a new `ApiResponse` object with the error details.
     * ***/
    suspend fun <T> serverRequestPatchAsync(
        context: Context, endpoint: String, referralParams: ReferralParams
    ): ApiResponse {

        val jsonMediaType = "application/json".toMediaTypeOrNull()
        val requestBody: RequestBody = Gson().toJson(referralParams).toRequestBody(jsonMediaType)
        val urlBuilder = (PrefHelper.aPIBaseUrl + endpoint).toHttpUrlOrNull()?.newBuilder()
        val url = urlBuilder?.build()?.toString()

        val requestBuilder =
            Request.Builder().url(url!!).addHeader("Authorization", RHUtil.readRhKey(context))
                .addHeader("Accept", "application/vnd.referralhero.v1").patch(requestBody)

        val request = requestBuilder.build()
        return withContext(Dispatchers.IO) {
            val response = client.newCall(request).execute()
            val responseString = response.body?.string()
            val parsedResponse: ApiResponse = gson.fromJson(responseString, ApiResponse::class.java)
            if (response.code == 200) {
                parsedResponse
            } else {
                ApiResponse("error", parsedResponse.message, parsedResponse.code, null, null, 0)
            }

        }
    }


}